File: tutorials/integration/01-schema-discovery/detect.py
#!/usr/bin/env python3
"""Discover SAF's API capabilities at runtime using schema().

Shows how an AI agent or tool integrator can programmatically
discover all available queries, selectors, and graph types.

Usage:
    python detect.py
"""

import subprocess
import json
from pathlib import Path

from saf import Project


def main() -> None:
    tutorial_dir = Path(__file__).parent
    source = tutorial_dir / "program.c"
    llvm_ir = tutorial_dir / "program.ll"

    subprocess.run(
        ["clang-18", "-S", "-emit-llvm", "-O0", "-g",
         "-o", str(llvm_ir), str(source)],
        check=True,
    )

    proj = Project.open(str(llvm_ir))

    # Discover all capabilities
    schema = proj.schema()

    print("SAF Schema Discovery")
    print("=" * 50)
    print(f"\nSchema keys: {list(schema.keys())}")

    for key, value in schema.items():
        print(f"\n--- {key} ---")
        if isinstance(value, dict):
            for k, v in value.items():
                print(f"  {k}: {v}")
        elif isinstance(value, list):
            for item in value:
                print(f"  - {item}")
        else:
            print(f"  {value}")

    # Pretty-print as JSON for external tools
    print("\n\nFull schema (JSON):")
    print(json.dumps(schema, indent=2, default=str))


if __name__ == "__main__":
    main()

File: tutorials/integration/01-schema-discovery/detect.rs
// Discover SAF's capabilities using the Rust API.
//
// The Rust API doesn't have a schema() equivalent, but you can
// inspect the module structure and available analysis types.
//
// Build and run:
//   cargo run --features llvm-18 --example detect_schema_discovery

use std::path::PathBuf;
use std::process::Command;

use saf_analysis::callgraph::CallGraph;
use saf_analysis::cfg::Cfg;
use saf_analysis::defuse::DefUseGraph;
use saf_core::config::Config;
use saf_frontends::api::Frontend;
use saf_frontends::llvm::LlvmFrontend;

fn main() {
    let tutorial_dir = PathBuf::from("tutorials/integration/01-schema-discovery");
    let source = tutorial_dir.join("program.c");
    let llvm_ir = tutorial_dir.join("program.ll");

    let status = Command::new("clang-18")
        .args(["-S", "-emit-llvm", "-O0", "-g",
               "-o", llvm_ir.to_str().unwrap(), source.to_str().unwrap()])
        .status()
        .expect("failed to run clang-18");
    assert!(status.success());

    let frontend = LlvmFrontend::new();
    let config = Config::default();
    let bundle = frontend.ingest(&[llvm_ir.as_path()], &config)
        .expect("failed to load LLVM IR");
    let module = bundle.module;

    println!("SAF Module Discovery:");
    println!("  Functions: {}", module.functions.len());
    println!("  Globals: {}", module.globals.len());

    // Available analysis types
    println!("\nAvailable analyses:");
    println!("  - CFG (control flow graph)");
    println!("  - CallGraph (inter-procedural call graph)");
    println!("  - DefUse (definition-use chains)");
    println!("  - PTA (points-to analysis)");
    println!("  - ValueFlow (value flow graph)");

    // Function details
    println!("\nFunction details:");
    for func in &module.functions {
        println!("  {} ({} blocks, {} params, is_decl={})",
                 func.name, func.blocks.len(), func.params.len(), func.is_declaration);
    }
}
